### ISSUE TITLE
Problems Getting TensorFlow to behave Deterministically

### ISSUE THREAD

**Issue by rundembear** [tensorflow/tensorflow#16889]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:

Yes, the code I am working with is company proprietary. If you need an example I will need to try to extract something that I can share. 

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:

Mac OS 10.13.2

- **TensorFlow installed from (source or binary)**:

binary

- **TensorFlow version (use command below)**:

v1.3.0-rc2-20-g0787eee 1.3.0

- **Python version**: 

3.6.4

- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:

Running on CPU.

- **GPU model and memory**:

- **Exact command to reproduce**:

Please see description of problem. I don't currently have an example I can give you but I am hoping that the information I am providing is enough to get at least some idea about what is going on.

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

I am having a major struggle trying to get TensorFlow to behave in a deterministic manner. I started out trying to compare two different versions of my source code, but the numeric differences from one execution to the next were large enough to make comparisons problematic.

So I went to trying to run exactly the same code in two different shells. And this is where things got weird.

I went through the code, and everywhere where TensorFlow is imported, everywhere I did operations using the form "with graph as default:", and prior to each call to run I added in a call to     tf.set_random_seed(42) [and yes, I used 42 in all of the calls].

I also placed breakpoints before anything in my code which was a tf call with 'random' in the name (I even searched the docs to see if there were random calls that didn't have random in the name, I found a couple but I am not using them). In all of the random calls that I am using I added in a seed=42 argument. So as far as I can tell I have covered all of the bases.

So what happens? I run the code in one shell, get a set of numbers. Run code in a second shell, get identical numbers. Good. Run the code a few more times, keep getting the same numbers. Better.
Run the code again, get completely different numbers!!! If I keep running the code in the different shells sometimes I get the same numbers from execution to execution. Sometimes I get different numbers. Sometimes it goes on to completely new numbers. Sometimes it goes back to an older set of numbers.

So I am getting something that is consistent enough that it feels like I haven't missed any places where I need to set the seed, but not consistent enough to say that it is deterministic.

I really want to be able to run the code over and over again and get exactly the same numeric result. Is this an unreasonable expectation? 

Can anyone offer an explanation as to why the code would generate exactly the same result (and I'm looking down to 8 decimal places so this is fairly precise) five times in a row, and then switch to a completely different result and then repeat that result multiple times in a row?

I can certainly try to reduce this into an example I can share (the model I am using is a toy model we use for unit testing). I wanted to see if anyone had any insight into the general problem before I put the time into that.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

I will include two examples of the numeric output I am getting. This data is part of the returned result from Session.run().

Example A:

out_node_weights [[ 9.09113979  9.09934711  9.09535408  9.09768963  9.11262989  9.09756184
   9.10322666  9.09527016  9.10811996  9.09029293]
 [ 9.34178257  9.34915924  9.34216785  9.34740925  9.32617569  9.34794617
   9.34459496  9.35203457  9.34202194  9.34494209]
 [ 9.09113979  9.09934711  9.09535408  9.09768963  9.11262989  9.09756184
   9.10322666  9.09527016  9.10811996  9.09029293]
 [ 9.09113979  9.09934711  9.09535408  9.09768963  9.11262989  9.09756184
   9.10322666  9.09527016  9.10811996  9.09029293]
 [ 9.59115124  9.59841442  9.59205914  9.59892082  9.58935356  9.59801197
   9.59655094  9.5970974   9.5989027   9.59136868]
 [ 9.39457417  9.39968395  9.39309788  9.4028101   9.40468502  9.39697838
   9.39671326  9.39805126  9.39696026  9.39565849]]

Example B:

out_node_weights [[ 9.04592514  9.05451775  9.05065632  9.05263042  9.06681919  9.05208111
   9.05834484  9.05059719  9.06315136  9.04525948]
 [ 9.25311852  9.26151276  9.25516319  9.25806046  9.23816681  9.26067734
   9.25860214  9.26416397  9.25700569  9.25596619]
 [ 9.27453995  9.28063679  9.27472687  9.28252983  9.28667641  9.27761269
   9.27913475  9.27859116  9.28033924  9.27534485]
 [ 9.27453995  9.28063679  9.27472687  9.28252983  9.28667641  9.27761269
   9.27913475  9.27859116  9.28033924  9.27534485]
 [ 9.50198936  9.51004601  9.50419044  9.50922203  9.50067139  9.50993729
   9.50946426  9.5085659   9.51255226  9.50196838]
 [ 9.27453995  9.28063679  9.27472687  9.28252983  9.28667641  9.27761269
   9.27913475  9.27859116  9.28033924  9.27534485]]

**Comment by shivaniag** [comment]
Do you mind sharing the code you are running here, it is really difficult to help without that information. Thanks!

**Comment by rundembear** [comment]
Thanks for taking a look. I can try to code up something that replicates the behaviour. Before I do that, can you please confirm whether or not Tensorflow is meant to be able to be used in a way where the numeric results are identical over an indefinite number of executions?

The best I have been able to get is within 4-5 decimal places.

I come from a computer graphics background and am used to being able to rerun very complicated renders or simulations and getting exactly the same result every time if I haven't changed any settings (that is, if I compare the resulting binary files they are identical). While the documentation seems to suggest that Tensorflow can provide this level of consistency, I haven't seen it.

If you expect that it is possible to run a graph today and get exactly the same result I got running the same graph last month, then I will try to put together an example which fails to achieve that goal. Then we can figure out if this is me failing to understand something, or a bug in Tensorflow.

Thanks!

**Comment by shivaniag** [comment]
@rundembear, I believe we do not guarantee this level of determinism with our current code. I would though suggest to use newer versions of tensorflow (I see you are using v1.3), which might give better results. 
@ebrevdo do you have any comment on this, I see you are tracking a similar issue#14663.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @ebrevdo, @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yoavz** [comment]
Have you tried setting your `inter_op_parallelism_threads` to 1? Some core TensorFlow operations on multi-core CPU will be non-deterministic by default, due to parallel execution being enabled by default. 

https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lucav76** [comment]
It would be great to be able to get a more deterministic behaviour, above all while using GPU training, as this would make hyper-parameters tuning much more useful and doable.
Nowadays it can really change quite much from one training to the next. In case that there is already a way to reduce that, I would appreciate it if you could provide some documentation.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jchia** [comment]
> Have you tried setting your `inter_op_parallelism_threads` to 1? Some core TensorFlow operations on multi-core CPU will be non-deterministic by default, due to parallel execution being enabled by default.
> 
> https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads

@yoavz Why would inter-op make cause non-deterministic results instead of intra-op. For intra-op, I expect that you can get different order of arithmetic operations in every run because of how a tensor operation is broken up to be run on different threads resulting in different floating-point error. For inter-op, I don't understand how.

**Comment by duncanriach** [comment]
I understand that this issue, as it started out, is related to determinism when running on a CPU, but you might be interested in the work that I've been doing on determinism in TensorFlow focused on GPUs.

For more information, see: https://github.com/NVIDIA/tensorflow-determinism

@lucav76 The reproducibility on GPUs that you have been wanting should now be available.

**Comment by kumariko** [comment]
@rundembear We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.6 which is latest stable version of TF and let us know if the issue still persists in newer versions. Thanks!
